cluster/addons/dns/kube2sky/kube2sky.go

```diff
@@ -70,43 +83,146 @@ type kube2sky struct {
-	// if PortalIP is not set, a DNS entry should not be created
-	if !kapi.IsServiceIPSet(service) {
-		glog.V(1).Infof("Skipping dns record for headless service: %s\n", service.Name)
+func (ks *kube2sky) writeSkyRecord(subdomain string, data string) error {
+	// Set with no TTL, and hope that kubernetes events are accurate.
+	_, err := ks.etcdClient.Set(skymsg.Path(subdomain), data, uint64(0))
+	return err
+}
+
+func (ks *kube2sky) deleteHeadlessService(svc *kapi.Service) {
+	ks.mlock.Lock()
+	defer ks.mlock.Unlock()
+	glog.V(5).Infof("deleting svc %q from headless service map", svc.Name)
+	delete(ks.headlessServicesMap, nameNamespace{svc.Namespace, svc.Name})
+}
+
+// Generates skydns records for a headless service.
+func (ks *kube2sky) newHeadlessService(subdomain string, service *kapi.Service) error {
+	// Create an A record for every pod in the service.
+	// This record must be periodically updated.
+	// Format is as follows:
+	// For a service x, with pods a and b create DNS records,
+	// a.x.ns.domain. and, b.x.ns.domain.
+	// TODO: Handle multi-port services.
+	ks.mlock.Lock()
+	defer ks.mlock.Unlock()
+	ks.headlessServicesMap[nameNamespace{service.Namespace, service.Name}] = service
+	key, err := kcache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		return err
+	}
+
+	e, exists, err := ks.endpointsStore.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to get endpoints object from endpoints store - %v", err)
+	}
+	if !exists {
+		glog.V(1).Infof("could not find endpoints for service %q in namespace %q", service.Name, service.Namespace)
 		return nil
 	}
+	if e, ok := e.(*kapi.Endpoints); ok {
+		return ks.generateRecordsForHeadlessService(subdomain, e, service)
+	}
+	return nil
+}
 
-	for i := range service.Spec.Ports {
-		svc := skymsg.Service{
-			Host:     service.Spec.PortalIP,
-			Port:     service.Spec.Ports[i].Port,
-			Priority: 10,
-			Weight:   10,
-			Ttl:      30,
+func getSkyMsg(ip string, port int) *skymsg.Service {
+	return &skymsg.Service{
+		Host:     ip,
+		Port:     port,
+		Priority: 10,
+		Weight:   10,
+		Ttl:      30,
+	}
+}
+
+func (ks *kube2sky) generateRecordsForHeadlessService(subdomain string, e *kapi.Endpoints, svc *kapi.Service) error {
+	for idx := range e.Subsets {
+		for subIdx := range e.Subsets[idx].Addresses {
+			subdomain := buildDNSNameString(subdomain, fmt.Sprintf("%d%d", idx, subIdx))
+			b, err := json.Marshal(getSkyMsg(e.Subsets[idx].Addresses[subIdx].IP, svc.Spec.Ports[0].Port))
+			if err != nil {
+				return err
+			}
+			glog.V(2).Infof("Setting DNS record: %v -> %q\n", subdomain, string(b))
+			if err := ks.writeSkyRecord(subdomain, string(b)); err != nil {
+				return err
+			}
 		}
-		b, err := json.Marshal(svc)
+	}
+
+	return nil
+
+}
+
+func (ks *kube2sky) addDNSUsingEndpoints(e *kapi.Endpoints) error {
+	ks.mlock.Lock()
+	defer ks.mlock.Unlock()
+	svc, exists := ks.headlessServicesMap[nameNamespace{e.Namespace, e.Name}]
+	if !exists {
+		return nil
```

Comment:
Thats true. I can use the service store instead of the additional map. I will make that change.